pkg/models/retry.go

```diff
@@ -40,68 +49,113 @@ func (t timerImpl) After(d time.Duration) <-chan time.Time {
+		retryIf:   func(err error) bool { return err != nil },      // retry on any error by default (legacy)
+		retryIfV2: func(resp *http.Response, err error) bool {
+			return err != nil || (resp != nil && resp.StatusCode >= http.StatusBadRequest)
+		}, // retry on any error or 4xx/5xx response by default (new)
+		timer:                 &timerImpl{},
+		context:               context.Background(),
+		returnHTTPStatusAsErr: true, // Legacy behavior: convert HTTP status to errors
 	}
 }
 
-// RetryableFuncWithResponse represents a function that returns an HTTP response or an error.
-type RetryableFuncWithResponse func() (*http.Response, error)
-
-// Retry retries the provided retryableFunc according to the retry configuration options.
-func Retry(retryableFunc RetryableFuncWithResponse, options ...RetryOption) (*http.Response, error) {
-	opts := newDefaultRetryConfig()
+func NewRetryConfig(options ...RetryOption) *RetryConfig {
+	opts := NewDefaultRetryConfig()
 
 	for _, opt := range options {
 		if opt != nil {
 			opt(opts)
 		}
 	}
+	return &RetryConfig{
+		retries:               opts.retries,
+		backoff:               opts.backoff,
+		maxJitter:             opts.maxJitter,
+		onRetry:               opts.onRetry,
+		onRetryV2:             opts.onRetryV2,
+		retryIf:               opts.retryIf,
+		retryIfV2:             opts.retryIfV2,
+		timer:                 opts.timer,
+		context:               opts.context,
+		returnHTTPStatusAsErr: opts.returnHTTPStatusAsErr,
+	}
+}
+
+// RetryableFuncWithResponse represents a function that returns an HTTP response or an error.
+type RetryableFuncWithResponse func() (*http.Response, error)
 
+// Retry retries the provided retryableFunc according to the retry configuration options.
+func Retry(retryableFunc RetryableFuncWithResponse, retryConfig *RetryConfig) (*http.Response, error) {
 	var lastErr error
-	for n := uint(0); n < opts.retries; n++ {
-		if err := opts.context.Err(); err != nil {
+	var lastResp *http.Response
+
+	for n := uint(0); n < retryConfig.retries; n++ {
+		if err := retryConfig.context.Err(); err != nil {
 			return nil, err
 		}
 
 		resp, err := retryableFunc()
+
+		// If the response is successful, return it immediately
 		if err == nil && resp != nil && resp.StatusCode == http.StatusOK {
 			return resp, nil
 		}
 
+		// Store the response and error for potential return
+		lastResp = resp
+		lastErr = err
+		statusAsErr := errors.New(resp.Status)
+		errResult := err
+
+		// Determine if we should retry based on error or response
+		retryIfV1Err := err
+		shouldRetry := false
 		if err == nil && resp != nil {
-			err = errors.New(resp.Status)
+			retryIfV1Err = statusAsErr
+			// set errResult and lastErr based on behavior flag
+			if retryConfig.returnHTTPStatusAsErr {
+				errResult = statusAsErr
+				lastErr = statusAsErr
+			}
 		}
 
-		if !opts.retryIf(err) {
-			return nil, err
+		shouldRetry = retryConfig.retryIf(retryIfV1Err) || retryConfig.retryIfV2(resp, err)
```

Comment:
Note: we call both the legacy and new retryIf for backwards compatibility purposes